spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/PushDown.scala

```diff
@@ -0,0 +1,204 @@
+
+  def isRangeComparison(filter: Filter) : Boolean = filter match {
+    case _: sources.LessThan           => true
+    case _: sources.LessThanOrEqual    => true
+    case _: sources.GreaterThan        => true
+    case _: sources.GreaterThanOrEqual => true
+    case _                             => false
+  }
+
+  def isSingleColumn(filter: Filter) = filter.isInstanceOf[sources.In] || filter.isInstanceOf[sources.EqualTo] ||
+    filter.isInstanceOf[sources.LessThan] || filter.isInstanceOf[sources.LessThanOrEqual] ||
+    filter.isInstanceOf[sources.GreaterThan] || filter.isInstanceOf[sources.GreaterThanOrEqual]
+
+
+  /** Returns the only column name referenced in the predicate */
+  def predicateColumnName(filter: Filter) : String = {
+    filter match {
+      case eq: sources.EqualTo            => eq.attribute
+      case lt: sources.LessThan           => lt.attribute
+      case le: sources.LessThanOrEqual    => le.attribute
+      case gt: sources.GreaterThan        => gt.attribute
+      case ge: sources.GreaterThanOrEqual => ge.attribute
+      case in: sources.In                 => in.attribute
+      case _ =>
+        throw new UnsupportedOperationException(
+          "It's not a valid filter to be pushed down, only >, <, >=, <= and In are allowed: " + filter)
+    }
+  }
+}
+
+/**
+ * Calculate the pushdown predicates for a given table and predicates.
+ *
+ * Partition pruning predicates are also detected an applied.
+ *  1. Only push down no-partition key column predicates with =, >, <, >=, <= predicate
+ *  2. Only push down primary key column predicates with = or IN predicate.
+ *  3. If there are regular columns in the pushdown predicates, they should have
+ *     at least one EQ expression on an indexed column and no IN predicates.
+ *  4. All partition column predicates must be included in the predicates to be pushed down,
+ *     only the last part of the partition key can be an IN predicate. For each partition column,
+ *     only one predicate is allowed.
+ *  5. For cluster column predicates, only last predicate can be non-EQ predicate
+ *     including IN predicate, and preceding column predicates must be EQ predicates.
+ *     If there is only one cluster column predicate, the predicates could be any non-IN predicate.
+ *  6. There is no pushdown predicates if there is any OR condition or NOT IN condition.
+ *  7. We're not allowed to push down multiple predicates for the same column if any of them
+ *     is equality or IN predicate.
+ *
+ */
+abstract class PushDown[T](predicates: Seq[T], tableDef: TableDef) {
+
+  def isEqualTo(predicate: T): Boolean
+
+  def isIn(predicate: T) : Boolean
+
+  def isRangeComparison(predicate: T) : Boolean
+
+  def isSingleColumn(predicate: T) : Boolean
+
+  /** Returns the only column name referenced in the predicate */
+  def predicateColumnName(predicate: T) : String
+
+  private val partitionKeyColumns = tableDef.partitionKey.map(_.columnName)
```

Comment:
Aren't column names provided by Spark SQL lowercased? It was mentioned somewhere earlier in this PR. If so, you should probably somehow take into account that fact because you are comparing the columns from schema with the columns from predicates.